File: Leetcode/ReverseLinkedList/Reverse_Linked_List.cpp
#include <bits/stdc++.h>
#include <list>
using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
 
struct ListNode {                                          
    int val;                                               
    ListNode *next;                                        
    ListNode() : val(0), next(nullptr) {}                  
    ListNode(int x) : val(x), next(nullptr) {}             
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
 
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
    	ListNode* ptr;
        list<int> L1;
        list<int>::iterator it;
        ptr=head;
        while(ptr!=NULL){
        	L1.push_back(ptr->val);
        	ptr=ptr->next;
		}
		L1.reverse();
		ptr=head;
		for(it=L1.begin();it!=L1.end();it++){
			ptr->val=*it;
			ptr=ptr->next;
		}
		return head;
    }
};

File: Leetcode/206.ReverseLinkedList/PYTHON/Reverse_Linked_List.py
# -*- coding: utf-8 -*-
"""
Created on Sat Apr 15 14:41:23 2023

@author: User
"""
from typing import Optional
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
         self.val = val
         self.next = next
class Solution:
    def recursive(self,cur :Optional[ListNode],ptr :Optional[ListNode])->Optional[ListNode]:
        if ptr==None:
            return cur
        temp=ptr.next
        pre=ptr
        pre.next=cur
        return self.recursive(pre,temp)
        
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        cur=None
        result=self.recursive(cur,head)
        return result
    
sol=Solution()
head=ListNode(1)
ptr=head
ptr.next=ListNode(2)
ptr.next.next=ListNode(3)
ptr.next.next.next=ListNode(4)
ptr.next.next.next.next=ListNode(5)
result=sol.reverseList(ptr)
while result!=None:
    print(result.val)
    result=result.next